pysindy/optimizers/sbr.py

```diff
@@ -76,22 +76,27 @@ def __init__(
         slab_shape_nu=4,
         slab_shape_s=2,
         noise_hyper_lambda=1,
-        normalize_columns=False,
-        copy_X=True,
-        **mcmc_kwargs,
+        num_warmup=1000,
+        num_samples=5000,
+        mcmc_kwargs=None,
+        **kwargs,
     ):
-        super().__init__(
-            copy_X=copy_X,
-            normalize_columns=normalize_columns,
-        )
+        super().__init__(**kwargs)
         # set the hyperparameters
         self.sparsity_coef_tau0 = sparsity_coef_tau0
         self.slab_shape_nu = slab_shape_nu
         self.slab_shape_s = slab_shape_s
         self.noise_hyper_lambda = noise_hyper_lambda
 
-        # store the MCMC kwargs.
-        self.mcmc_kwargs = mcmc_kwargs
+        # set MCMC sampling parameters.
+        self.num_warmup = num_warmup
+        self.num_samples = num_samples
+
+        # set the MCMC kwargs.
+        if mcmc_kwargs is not None:
+            self.mcmc_kwargs = mcmc_kwargs
+        else:
+            self.mcmc_kwargs = {}
 
     def _reduce(self, x, y):
```

Comment:
Is `y` is expected to be measurements of the system, in [this line](https://github.com/dynamicslab/pysindy/blob/01698bee0a6ed69209d132fca065b6f0ec68a89d/pysindy/optimizers/sbr.py#L128)?  In general SBR, as a SINDy optimizer, the `y` argument to `_reduce` will receive the estimated `x_dot` from whatever differentiation method is used.  Callers can skip the differentiation step by directly passing derivatives, e.g. `SINDy(optimizer=SBR()).fit(x=x, x_dot=x)`.  If I'm understanding this correctly, we should mention that in the docstring.

Long term, I have a `sss` branch in which I hope to implement a more clear `SINDy` subclass for these kinds of problems that eschew a differentiation method.

Reply:
@Jacob-Stevens-Haas that's right, `y` is expecting derivatives `x_dot`, like the other optimizers.

The line `numpyro.sample("obs", Normal(mu, sigma), obs=y)` is the likelihood, and it says that the errors of these derivatives are Gaussian. It's maybe a little misleading that the parameter of the likelihood is called `obs`, but that's kind of a quirk the Bayesian formalism.

Reply:
@Jacob-Stevens-Haas I think I might have made a massive blunder and overlooked something in Hirsh's paper.

The observation model and likelihood is with respect to the states in Hirsh, not the derivatives. The way I have implemented it, the likelihood is with respect to the derivatives.

I don't think it will matter much in terms of the actual numerical results for the coefficients, especially since the noise parameter `sigma` is free. However, the current code does not do what is described in Hirsh's paper.

I think what we would actually want is to be able to reconstruct the states from within the `_numpyro_model`, basically by running something equivalent of `pysindy.simulate`, i.e.

I don't know what is the best thing to do here, I'll have to think this through.

```
# sample the parameters compute the predicted outputs.
beta = _sample_reg_horseshoe(tau, c_sq, (n_targets, n_features))

# get SINDy state prediction given coefficients beta
x_hat = model.simulate(beta)

# compute the likelihood.
sigma = numpyro.sample("sigma", Exponential(self.noise_hyper_lambda))
numpyro.sample("obs", Normal(x_hat, sigma), obs=y)
```

Reply:
I wouldn't worry too much.  There's a few other places in pysindy that the choice of one component requires the user to handle things differently outside of that component: `TrappingSR3` optimizer requires the user to choose a degree-2 polynomial library, and `WeakPDELibrary` means that the `SINDy` the differentiation method ends up as a no-op.  In the latter case, we wrote a hook for feature libraries to control how differentiation is applied, but that's little more than a one-time hack that we can't do here.

Short term, we'll just have to expect user to know to only use `SBR` in cases where `model.fit(x=x, x_dot=x)`.  I think the PR has enough merit to users, since users of bleeding-edge SINDy are almost exclusively researching the methods themselves, rather than applying them to real world data.  If it's restrictions are clear in the docstring, I'm happy.

Long term, this is discussed in #351 and #407, which I'll start working on once I've cleaned up the trapping PRs.  This PR is a great template for how they should be implemented.

Reply:
Ohh, I see I misunderstood the role of `mu = jnp.dot(x, beta.T)`.  Yeah, the paper does simulate.  Maybe in place of 3.3/3.4 we use the random variable that equates to standard SINDy, i.e. $\dot x \sim$ Normal $(\Xi^T\Theta(x), \sigma))$, but keep the prior $\sigma\sim$ Exp $(\lambda)$?

Reply:
@Jacob-Stevens-Haas Thanks for looking into this.

I have been doing a bit of digging, and it will be significantly more complex to do what is in the paper. Mainly because `jax` tracks operations to compute derivatives. This is not compatible with e.g. `pysindy.feature_library` (which we would need for the `predict` step in the simulation), where e.g. `np.asarray` is employed for type coercion.

Similarly, `scipy.integrate.solve_ivp` also employs type coercion, but to use the NUTS sampler in `numpyro`, gradients would have to be tracked by `jax` through the numerical simulation. There is a `jax`-compatible ODE solver called [`diffrax`](https://github.com/patrick-kidger/diffrax) which could be employed to this end, but that is a relatively complex software engineering task.

I think the current implementation is valid for `discrete_time` simulations, which we could make clear in the documentation. I would be interested in looking into a continuous-time implementation too, using e.g. `diffrax` for simulating the forward model, but I think that's a bit of a longer term goal.

Reply:
Ok, thanks.  Yeah, long term diffrax would be great to have, as would removing all calls to `np.asarray`.  Relatedly, this repo also maintains a subclass of numpy array (See #451) to allow axis labels to propagate through algorithms, which is super useful in debugging, but it's stripped whenever we use `np.asarray`

Reply:
> Relatedly, this repo also maintains a subclass of numpy array (See https://github.com/dynamicslab/pysindy/pull/451) to allow axis labels to propagate through algorithms,

Yes, I have noticed that. Do you know of [Xarray](https://xarray.dev/)? I think they are doing something similar.